comms/src/control_service/worker.rs

```diff
@@ -0,0 +1,220 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::{
+    error::ControlServiceError,
+    service::ControlServiceConfig,
+    types::{ControlMessage, ControlServiceDispatcher, ControlServiceMessageContext, Result},
+};
+
+use crate::{
+    connection::{
+        connection::EstablishedConnection,
+        monitor::{ConnectionMonitor, SocketEvent, SocketEventType},
+        Connection,
+        Context,
+        Direction,
+        InprocAddress,
+    },
+    dispatcher::{DispatchResolver, DispatchableKey},
+    message::{FrameSet, Message, MessageEnvelope},
+    types::MessageEnvelopeHeader,
+};
+use std::{
+    convert::TryInto,
+    sync::mpsc::{sync_channel, Receiver, SyncSender},
+    thread,
+    time::Duration,
+};
+
+const LOG_TARGET: &'static str = "comms::control_service::worker";
+const CONTROL_SERVICE_MAX_MSG_SIZE: u64 = 300; // 1kb
+
+/// The [ControlService] worker is responsible for handling incoming messages
+/// to the control port and dispatching them using the message dispatcher.
+pub struct ControlServiceWorker<MType, R>
+where
```

Comment:
Is this required?
Generally you only want to define them at impl level

Reply:
Good catch - the `K` is required because I'm storing a hashmap in the dispatcher (so it needs `Eq+ Hash`) but the resolver bound is not needed - fixed